[PrintAsObjC] Remove empty extensions from Swift compatibility headers.

calebd opened the pull request:
This change removes empty extension interfaces from generated Swift compatibility headers where empty extensions add neither public members or protocol conformances.

Resolves [SR-2772](https://bugs.swift.org/browse/SR-2772).

calebd committed:
Add failing tests.

calebd committed:
Remove empty extensions from Swift compatibility headers.

calebd committed:
Restore empty extensions when they add protocol conformances.

jrose-apple commented:
This is definitely an improvement over the status quo, but it's not quite the whole SR-2772. Note that in that example there *is* a member in the extension; it's just not one that's printed. There could be a variety of reasons for this, the top two being access control and whether it's actually ObjC-compatible.

calebd committed:
Add a few more failing tests.

- Extension that only contains non-objc-compatible members
- Extensions that only add non-objc-compatible protocol conformances

calebd committed:
Change conditions under which an extension is considered empty.

- Empty extensions add no Objective-C compatible members or protocols

jrose-apple commented on test/PrintAsObjC/extensions.swift:
```diff
@@ -95,8 +118,7 @@ extension GenericClass {
+// NEGATIVE-NOT: @interface NSObject (SWIFT_EXTENSION(extensions))
 // NEGATIVE-NOT: @interface NSObject{{$}}
 // NEGATIVE-NOT: @class NSObject
 extension NSObject {}
```
This part is supposed to be a minimal test that we don't try to *define* an imported class. Can you put a member in instead, like you did with A2?

calebd replied:
Makes sense. In `master` it looks like the test first asserts that there is an extension, then asserts that there is no class defined. Due to the sorted nature of this header would it be more correct to first assert that the class declaration is not there followed by the extension check?

jrose-apple replied:
The CHECK and NEGATIVE lines are being checked in two different passes, so it's not going to be a problem in practice. I'm happy to have them reordered, though; that would match the cases where we *do* expect the extended class to be defined first.

jrose-apple commented on lib/PrintAsObjC/PrintAsObjC.cpp:
```diff
@@ -329,7 +329,33 @@ class ObjCPrinter : private DeclVisitor<ObjCPrinter>,
+  bool isEmptyExtensionDecl(ExtensionDecl *ED) {
+    auto members = ED->getMembers();
+    SmallVector<Decl *, 4> includedMembers;
+    std::copy_if(members.begin(), members.end(),
```
You can do even better by using [`std::any_of`](http://en.cppreference.com/w/cpp/algorithm/all_any_none_of).

jrose-apple reviewed (commented):
Thanks! This is more what we want. :-) A few comments, still.

calebd committed:
Switch from std::copy_if to std::any_of.

calebd committed:
Remove redundant protocol conformance test.

This test was designed to ensure that extensions that only add a
non-objc-compatible protocol are not printed in a compatibility header.
That case is already covered by the `extension NSString : NotObjC {}`
test in `PrintAsObjC/protocols.swift`.

calebd committed:
Restore test to ensure an extended imported class is not redefined.

jrose-apple commented:
@swift-ci Please test

jrose-apple commented:
Merged. Thanks, Caleb!
